frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -438,6 +440,33 @@ int SENTINEL2Dataset::Identify( GDALOpenInfo *poOpenInfo )
         strstr(pszHeader, "User_Product_Level-2A" ) != nullptr )
         return TRUE;
 
+    /* Last effort – look inside of a zip file for metadata file with
+     * matching name. */
+    if ( EQUAL(CPLGetExtension(pszJustFilename), "zip") )
+    {
+        CPLString osFilename(poOpenInfo->pszFilename);
+        if( strncmp(osFilename, "/vsizip/", strlen("/vsizip/")) != 0 )
+            osFilename = "/vsizip/" + osFilename;
+
+        auto psDir = VSIOpenDir(osFilename.c_str(), 1, nullptr);
+        while ( const VSIDIREntry* psEntry = VSIGetNextDirEntry(psDir) )
+        {
+            const char* pszInsideFilename = CPLGetFilename(psEntry->pszName);
+            if ( VSI_ISREG(psEntry->nMode) && (
+                STARTS_WITH_CI(pszInsideFilename, "MTD_MSIL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "MTD_MSIL2A") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_OPER_MTD_SAFL1B") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_OPER_MTD_SAFL1B") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_OPER_MTD_SAFL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_OPER_MTD_SAFL1C") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2A_USER_MTD_SAFL2A") ||
+                STARTS_WITH_CI(pszInsideFilename, "S2B_USER_MTD_SAFL2A")
```

Comment:
To avoid copy&paste, that list of filename prefixes should be moved to a dedicated function, that could be used by Open() as well

Reply:
Done. Is it good enough?